Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+        return try self.set(bytes: UnsafeRawBufferPointer(start: string.utf8Start, count: string.utf8CodeUnitCount), at: index)
+    }
+    
+    // MARK: String APIs
+    
+    /// Write `string` into this `SimpleRingBuffer` using UTF-8 encoding, moving the writer index forward appropriately.
+    ///
+    /// - parameters:
+    ///     - string: The string to write.
+    /// - returns: The number of bytes written.
+    @discardableResult
+    public mutating func write(string: String) throws -> Int {
+        let written = try self.set(string: string, at: self.ringTail)
+        self._moveTail(forwardBy: written)
+        return written
+    }
+    
+    /// Write `string` into this `SimpleRingBuffer` at `index` using UTF-8 encoding. Does not move the writer index.
+    ///
+    /// - parameters:
+    ///     - string: The string to write.
+    ///     - index: The index for the first serialized byte.
+    /// - returns: The number of bytes written.
+    @discardableResult
+    internal mutating func set(string: String, at index: Int) throws -> Int {
+        return try self.set(bytes: string.utf8, at: index)
+    }
+    
+    /// Get the string at `index` from this `SimpleRingBuffer` decoding using the UTF-8 encoding. Does not move the reader index.
+    ///
+    /// - parameters:
+    ///     - index: The starting index into `ByteBuffer` containing the string of interest.
+    ///     - length: The number of bytes making up the string.
+    /// - returns: A `String` value deserialized from this `SimpleRingBuffer` or `nil` if the requested bytes aren't contained in this `ByteBuffer`.
+    internal func getString(at index: Int, length: Int) -> String? {
+        precondition(index >= 0, "index must not be negative")
+        precondition(length >= 0, "length must not be negative")
+        
+        guard length <= self.capacity else {
+            return nil
+        }
+        
+        return self.withVeryUnsafeBytes { ptr in
+            if self.capacity - index >= length {
+                // single read
+                return String(decoding: UnsafeBufferPointer(start: ptr.baseAddress!.advanced(by: index).assumingMemoryBound(to: UInt8.self), count: length), as: UTF8.self)
+            } else {
+                // double read-- we will need to provide an intermediate buffer for the String API to read.
+                let firstPart = self.capacity - index
+                var bytes = ContiguousArray<UInt8>()
+                bytes.reserveCapacity(length)
+                bytes.append(contentsOf: UnsafeBufferPointer(start: ptr.baseAddress!.advanced(by: index).assumingMemoryBound(to: UInt8.self), count: firstPart))
+                bytes.append(contentsOf: UnsafeBufferPointer(start: ptr.baseAddress!.assumingMemoryBound(to: UInt8.self), count: length - firstPart))
+                return String(decoding: bytes, as: UTF8.self)
+            }
+        }
+    }
+    
+    public func peekString(length: Int) -> String? {
```

Comment:
I think there's an oversupply of APIs here.

Let's again try not to over engineer this data structure. `NIOHPACK` shouldn't be a package that is depended on only to gain access to the ring buffer data structure. For the moment, the ring buffer should be treated as a purely internal implementation detail of `NIOHPACK`. If it becomes broadly useful we can always bring it up to the top level and make it more general, but for now I'm inclined to want to give it enough API surface to support what we need it to do and no more.

My gut instinct is that if the API is only ever called by test code, then unless it serves a very specific diagnostic debugging purpose that cannot be replicated outside of the module it should be removed, along with the test that uses it.